### ISSUE TITLE
Running batch norm updates

### ISSUE THREAD

**Issue by ygoncharov** [uber/horovod#49]
Is there a simple way to run batch norm updates in addition to gradient updates? The industry practice is to run batch update ops from a single tower only, therefore, it would be incorrect to add batch updates to the training op in the provided sample code.

**Comment by alsrgv** [comment]
I believe if you add batch normalization to the `train_op` following example [here](https://www.tensorflow.org/api_docs/python/tf/layers/batch_normalization), the batch norm would run on every worker independently, without exchanging or averaging those updates. Do you anticipate that creating any problems?

**Comment by ygoncharov** [comment]
As long as batch statistics is not aggregated, my understanding is that it should work for the regular batch norm. However, it may have some side effects for the batch renormalization since a tower won't have access correct batch statistics (i. e. the one used for inference) during training.

Also, we should keep in mind that we may get different inference results from each tower.

When a model is saved in the sample script, will it contain the batch statistics from the first tower only?

**Comment by alsrgv** [comment]
Great, that matches my understanding of regular batch normalization.

Can you elaborate a bit more on batch renormalization issue? Is it for inference or for training?

That's true that every tower will have a slightly different result. The example script right now is not configured to save checkpoint, but it can be easily done with hooks for [TF](https://www.tensorflow.org/versions/r1.4/api_docs/python/tf/train/CheckpointSaverHook) or [Keras](https://keras.io/callbacks/#example-model-checkpoints). One can check for `hvd.rank() == 0` and only add checkpoint hook on worker 0.

**Comment by ygoncharov** [comment]
For the batch renormalization, the activations computed in the forward pass of the training step are supposed to be identical to the activations computed in inference. For the proposed solution, it is not a case since only one worker has access to the batch statistics that will be used during inference. It may still successfully solve the problems the batch renormalization is supposed to solve (i. e. training with small mini-batch size), but it will be a bit different from the original algorithm.

**Comment by alsrgv** [comment]
I see, thanks for the clarification. It would be interesting to run this on few practical examples and see how much different those BN parameters would turn out to be.

**Comment by alsrgv** [comment]
I'll close this issue. Please feel free to re-open if you have additional questions or ideas.
